app/src/main/java/org/wikipedia/views/UserMentionEditText.kt

```diff
@@ -0,0 +1,197 @@
+                onCancelUserNameEntry()
+                return
+            }
+            onUserNameChanged(text!!.substring(userNameStartPos, userNameEndPos))
+        } else if (selStart == selEnd && !isEnteringUserName) {
+            val spans = text!!.getSpans(selStart, selEnd, UserColorSpan::class.java)
+            if (spans.isNotEmpty()) {
+                userNameStartPos = text!!.getSpanStart(spans[0])
+                userNameEndPos = text!!.getSpanEnd(spans[0])
+                onStartUserNameEntry()
+                onUserNameChanged(text!!.substring(userNameStartPos, userNameEndPos))
+            }
+        }
+    }
+
+    private fun onStartUserNameEntry() {
+        listener?.onStartUserNameEntry()
+    }
+
+    private fun onCancelUserNameEntry() {
+        userNameStartPos = -1
+        userNameEndPos = -1
+        listener?.onCancelUserNameEntry()
+    }
+
+    private fun onUserNameChanged(userName: String) {
+        L.d("Entering username: $userName")
+        listener?.onUserNameChanged(userName)
+    }
+
+    fun prepopulateUserName(userName: String) {
+        val sb = SpannableStringBuilder()
+        sb.append("@$userName")
+        val spanEnd = sb.length
+        sb.append(" ")
+        createUserNameSpan(sb, 0, spanEnd)
+        isUserNameCommitting = true
+        text = sb
+        isUserNameCommitting = false
+        setSelection(sb.length)
+    }
+
+    fun onCommitUserName(userName: String) {
+        try {
+            isUserNameCommitting = true
+            if (userNameStartPos < 0 || userNameEndPos <= userNameStartPos) {
+                onCancelUserNameEntry()
+                return
+            }
+
+            val sb = SpannableStringBuilder()
+            sb.append(text!!.subSequence(0, userNameStartPos))
+            val spanStart = sb.length
+            sb.append("@$userName")
+            val spanEnd = sb.length
+            if (userNameEndPos < text!!.length) {
+                sb.append(text!!.subSequence(userNameEndPos, text!!.length - 1))
+            }
+
+            createUserNameSpan(sb, spanStart, spanEnd)
+            text = sb
+            setSelection(spanEnd)
+            onCancelUserNameEntry()
+        } finally {
+            isUserNameCommitting = false
+        }
+    }
+
+    private fun createUserNameSpan(spannable: Spannable, start: Int, end: Int) {
+        val span = UserColorSpan(ResourceUtil.getThemedColor(context, R.attr.colorAccent))
+        spannable.setSpan(span, start, end, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
+    }
+
+    /**
+     * Returns the text of this field, with properly expanded user mentions.
+     * Each user mention (in th form of @{username}) will be expanded into a wiki link to the
+     * user page of that user, i.e. [[User:username|@username]]
+     */
+    fun getParsedText(wikiSite: WikiSite): String {
+        if (text == null) {
+            return ""
+        }
+        var str = text!!.toString()
```

Comment:
Would it be possible to assign a value of `text` so that we can reduce the non-null symbol on it? e.g.: `val editText = text!!`

Reply:
yep!